### DISCUSSION TITLE
Error

### DISCUSSION THREAD

**Discussion by matthewhisle** [ujjwalguptaofficial/JsStore#345] [General]
I am getting the following error... It ONLY happens after device has been used for a long time. If I refresh chrome cache, it works fine again

Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'name') at nt (jsstore.worker.min.js:2:64964) at jsstore.worker.min.js:2:69300

Again, if I refresh chrome and delete all data, it works again.

It appears that it has an issue with the dbName... AKA the name of the database. Any advice?

**Comment by matthewhisle** [comment]
const isDbCreated = await connection.initDb(database);
    if (isDbCreated === true) {
        //alert("db created");
    }
    else {
        // alert("db opened");
    }

It is failing on the line const isDbCreated


Without ANY code change, if I clear the chrome storage on phone, it starts to work again. This error is random, and only every once in awhile after lots of usage.

**Reply by ujjwalguptaofficial** [reply]
which version you are currently using ?

**Reply by ujjwalguptaofficial** [reply]
please point me to the code where the error is occurring, like highlighting the point please

**Reply by matthewhisle** [reply]
Thank you for the reply! I am currently using the LATEST version of JSSTORE.

![image](https://github.com/ujjwalguptaofficial/JsStore/assets/44656142/de3a4501-86fd-4513-8e25-97e385a05722)

It was failing right here on e.name .... This is in jsstore.worker.min.js

**Reply by ujjwalguptaofficial** [reply]
I will check this, thanks.

**Reply by matthewhisle** [reply]
Thank you! I really appreciate it. This will help me tremendously, we love jsstore!

**Reply by matthewhisle** [reply]
I saw an error before in this forum where someone got the exact same error that I got here. Then it was because the person wasn't initializing the connection right. Is it possible that the real error is in the initialization of the connection, then causing this error message to occur for me. Upon investigation, the connection is getting stuck and requests add up in the queue. Perhaps an error in the service worker?

**Reply by matthewhisle** [reply]
Thread name "connection.dropDb() not works" in these forums

You'll notice my error message is IDENTICAL to his. Except his was thrown simply because he did not initialize database first. Mine was thrown ON the database initialization call.

**Comment by matthewhisle** [comment]
Keep in mind, the oddity is that with this... It works 100% fine, then after awhile it starts to fail and gives the error up there on this line with 0 code changing... Then if I clear all chrome data, it works again

**Reply by ujjwalguptaofficial** [reply]
Thanks for the explanation, i will spend time on this and let you know.

**Reply by matthewhisle** [reply]
Thank you so much. Let me know if there is any way I can help.

**Comment by ujjwalguptaofficial** [comment]
![image](https://github.com/ujjwalguptaofficial/JsStore/assets/10938615/46720dd0-500e-4d61-be98-d21c15087343)

This is the code where error is occuring. 

but it's not possible as there is already a null check when this class is initiated .

![image](https://github.com/ujjwalguptaofficial/JsStore/assets/10938615/6fcc8491-f0b5-4bbe-bbda-fd37f489ef03)

I will keep my eye on it, but if you find something let me know.

**Reply by matthewhisle** [reply]
How do I check my version number to make sure you and I are looking at same version? I know I said I was on newest version, but I want to make sure I am not assuming to not mislead us into looking at different versions. Thank you!

**Reply by ujjwalguptaofficial** [reply]
check package.json

**Reply by matthewhisle** [reply]
Was there any recent version where you did not have a null check there? Because that is exactly where the error was getting thrown. And this was in the initialization of the database

**Reply by matthewhisle** [reply]
This is still happening. Not sure why.

**Reply by matthewhisle** [reply]
Same error message

**Reply by matthewhisle** [reply]
I think the error message is a RESULT of the connection var being messed up.

**Reply by matthewhisle** [reply]
Inactivity timer gets set to -1 and it starts to queue any task I try to run on the DB and won't run anything. The "request queue" just starts to pile up. It is like the worker is getting stuck somehow.

**Comment by matthewhisle** [comment]
It seems like its the connection var getting stuck. How often do connections expire? Do you recommend me wrapping the DB Initialization in a try catch and if it hits the catch, destroy the connection and recreate it? I am declaring the connection in a .js file that is used by multiple HTML pages.

**Reply by ujjwalguptaofficial** [reply]
It might be, but i am not able to reproduce.

```
Current process is when the request queue is empty, the connection closes the db and when there is some resource it's open the db and then process other requests.
```

You are onto something, please see if this is reproducible and tell me the steps.

**Reply by ujjwalguptaofficial** [reply]
You can turn on log and then figure out the request queue -

```
https://jsstore.net/docs/helpers#logstatus
```

After turning on the log, all request will be logged with start and finish.

**Reply by matthewhisle** [reply]
So, more info. I used the connection.terminate() for the first time....

What it is behaving like, is when this happens, the connection looks IDENTICAL to when you run connection.terminate(). The difference is, when I run connection.terminate(), if I refresh page (aka it reinitializes the connection), it is fine and back to normal. In this case it's like the connection gets permanently stuck in a terminated state. The var will exist, but it is not opened upon initialization. Any thoughts?

**Reply by ujjwalguptaofficial** [reply]
connection.terminate is supposed to terminate everything and release the resources. So it's normal - if you want to use query on this connection, you gotta reinitialize using `initDb` api.

**Reply by ujjwalguptaofficial** [reply]
https://jsstore.net/docs/terminate

**Reply by matthewhisle** [reply]
Understood. What I am referring to is when this glitch occurs, I am initializing the connection but it's stuck in terminated state with inactivity timer set to -1000

**Reply by ujjwalguptaofficial** [reply]
the reintialization should work, if its not working then please give me a reproducible repo.

Also if you are able to reproduce the issue, we can get on a call to debug and fix this.

**Reply by matthewhisle** [reply]
Thank you so much! Will do. Waiting to get my hands on a device of my team when it stops working. Will hit you up then!

**Comment by matthewhisle** [comment]
![image](https://github.com/ujjwalguptaofficial/JsStore/assets/44656142/09de8ebd-c844-419d-b837-d8aef85861b5)

**Reply by matthewhisle** [reply]
![image](https://github.com/ujjwalguptaofficial/JsStore/assets/44656142/51498f92-1a79-4bac-8b06-bb4fb197e1b7)

Finally got my hands on another device that failed. Note this only happens once every blue moon!

**Reply by matthewhisle** [reply]
![image](https://github.com/ujjwalguptaofficial/JsStore/assets/44656142/4f5a7842-8734-4c0e-a5ee-e5dcaa3d712b)

Init_DB itself gets stuck in the request queue

**Reply by matthewhisle** [reply]
Please let me know any additional info you need to debug. I am remote debugging the specific device that is "Locked Up"

**Comment by matthewhisle** [comment]
![image](https://github.com/ujjwalguptaofficial/JsStore/assets/44656142/f7305acc-486f-408f-b3fd-fd4ffc72c742)

Since its the connection itself seemingly getting stuck I can't even enable log status

**Reply by ujjwalguptaofficial** [reply]
Thanks, i will have a look. Please allow me sometime.

**Reply by ujjwalguptaofficial** [reply]
It's getting stucked because something is failing in the worker and that's why its not moving forward. I will have to see why initdb in the worker is failing.

**Reply by ujjwalguptaofficial** [reply]
could you tell me if database was created or opened ? @matthewhisle

**Reply by ujjwalguptaofficial** [reply]
also is it happening in mobile only or desktop also ?

**Reply by ujjwalguptaofficial** [reply]
Are you using multiple database or multiple connection ?

**Reply by matthewhisle** [reply]
The database was OPENED in this case not created. It is failing on reopening it. 1 database, 1 connection. It is a VERY hard issue to replicate. I have 50 devices deployed and only once per week it happens. No pattern in device, timing, etc. Completely random.

**Reply by matthewhisle** [reply]
100% agree. Something in worker is for sure failing on init DB.

**Reply by matthewhisle** [reply]
Right now I have to have my client clear all chrome cache to get it working again. Is there any programmatic way around this for the time being? Will connection.abort work or will it get stuck in the queue too? I don't want to fix this device too quickly in case I need to get you any more info off  of it.

**Reply by ujjwalguptaofficial** [reply]
I have released  a new version - 4.6.4. The error will be propagating if some error will occur in `initDb` - you can catch the error and then reinitiate database basically calling initDb.

**Reply by matthewhisle** [reply]
Sorry for my lack of understanding and thank you for this. Can you lay out the high level code I should use to make this happen?

**Reply by ujjwalguptaofficial** [reply]
I think try to reinitiate few times and then if it does not occurs then reload the page - 

let maxInitDbTryCount =2;

```
async function initiateDatabase() {
    try {
        await con.initDb(db);
    } catch (ex) {
        console.error("db failed");
        if (maxInitDbTryCount > 0) {
            maxInitDbTryCount--;
            initiateDatabase();
        } else {
            window.location.reload();
        }
    }
}
```

I hope this helps you.

**Reply by matthewhisle** [reply]
So, really weird thing. I try reinitializing 4 times, and even reload page, and the connection is still stuck in the same state.

**Reply by matthewhisle** [reply]
All the initializations get stuck in the request queue

**Reply by matthewhisle** [reply]
Then when refresh page, initialize, it still gets stuck in request queue

**Reply by ujjwalguptaofficial** [reply]
did you install the new version ?

**Reply by matthewhisle** [reply]
Whoops I am so sorry. I thought the new version just is able to catch error, does the new version have code that might fix?

**Reply by ujjwalguptaofficial** [reply]
could you give me url to your app ? I feel something is wrong in the implementation.

**Reply by matthewhisle** [reply]
I will have to push to my production site when I can if so. I only have one device that is stuck in this state and it is on my production site and it's hard for me to replicate error on another device

**Reply by matthewhisle** [reply]
Can I email you the code to the page

**Reply by matthewhisle** [reply]
Just emailed you the exact implementation code

**Reply by ujjwalguptaofficial** [reply]
I have added few things there to debug error and also propagate error till the request, so that connection is not getting stuck.

**Reply by matthewhisle** [reply]
![image](https://github.com/ujjwalguptaofficial/JsStore/assets/44656142/a16120f4-b7e0-410d-b9cd-d773710c778e)

Can I get the latest version from here

**Reply by ujjwalguptaofficial** [reply]
please update the version and see if its happening again.

**Reply by matthewhisle** [reply]
Can I get the latest one you just released from the CDN links there

**Reply by ujjwalguptaofficial** [reply]
Are you not using npm ? yeah you can get it from cdn also.

**Reply by matthewhisle** [reply]
I am not I am using cdn right now. Are there any benefits to npm other than the event of potentially a cdn link failing? Sorry I am mostly a SQL developer that uses javascript just enough to get the job done so I am always learning

**Reply by ujjwalguptaofficial** [reply]
yeah cdn is fine, no issue.

**Reply by ujjwalguptaofficial** [reply]
you can download from here also - https://github.com/ujjwalguptaofficial/JsStore/releases/tag/v4.6.4

**Reply by matthewhisle** [reply]
Oddly it isnt catching the error

**Reply by matthewhisle** [reply]
Hang on it may be my fault

**Reply by matthewhisle** [reply]
Even terminate gets stuck in the request queue as well

**Reply by ujjwalguptaofficial** [reply]
why you are calling terminate ?

**Reply by matthewhisle** [reply]
I will give this a try tomorrow when I am able to update my production site with this new code and will let you know what happens. Thank you!

**Reply by matthewhisle** [reply]
Just as a side note, I get the same error even if I initialize connection with no worker TypeError: Cannot read properties of undefined (reading 'name')
    at nt (<anonymous>:7:64964)
    at <anonymous>:7:69300

**Reply by matthewhisle** [reply]
Just learned something REALLY odd. If I initialize another database on page load with a different name it loads fine and now shows 2 databases in my application... So its like something is stuck with that SPECIFIC database name

![image](https://github.com/ujjwalguptaofficial/JsStore/assets/44656142/32169360-9004-42e1-b0b2-1db25f4e63bb)

**Reply by matthewhisle** [reply]
So if I try to run initDb on cadmusscribe it gives all those errors. If I run it to create cadmusscribe1, no issues. Both have identical schema

**Reply by matthewhisle** [reply]
I used indexedDB.deleteDatabase() on cadmusscribe, and now everything works great.

**Comment by matthewhisle** [comment]
Hey I have been using the indexeddb.deletedatabase() call on the catch to rebuild teh database when this happens and its been working okay for me, just wanted to check if you have seen any root cause for why the JSMeta is getting deleted and messing up the local database?
